core/src/main/java/org/apache/iceberg/util/ExpireSnapshotUtil.java

```diff
@@ -0,0 +1,397 @@
+                  // this commit ensures that the full ancestor list between when the snapshot was
+                  // written and this expiration is known and there is no missing history. If history
+                  // were missing, then the snapshot could be an ancestor of the table state but the
+                  // ancestor ID set would not contain it and this would be unsafe.
+                  manifestsToRevert.add(manifest.copy());
+                }
+              }
+            }
+            // add the manifest list to the delete set, if present
+            if (snapshot.manifestListLocation() != null) {
+              manifestListsToDelete.add(snapshot.manifestListLocation());
+            }
+          } catch (IOException e) {
+            throw new UncheckedIOException(
+                String.format("Failed to close manifest list: %s", snapshot.manifestListLocation()),
+                e);
+          }
+        });
+    return new ManifestExpirationChanges(manifestsToScan, manifestsToRevert, manifestsToDelete,
+        manifestListsToDelete);
+  }
+
+  private static final Schema MANIFEST_PROJECTION = ManifestFile.schema()
+      .select("manifest_path", "added_snapshot_id", "deleted_data_files_count");
+
+  private static CloseableIterable<ManifestFile> readManifestFiles(
+      Snapshot snapshot, TableOperations ops) {
+
+    if (snapshot.manifestListLocation() != null) {
+      return Avro.read(ops.io().newInputFile(snapshot.manifestListLocation()))
+          .rename("manifest_file", GenericManifestFile.class.getName())
+          .classLoader(GenericManifestFile.class.getClassLoader())
+          .project(MANIFEST_PROJECTION)
+          .reuseContainers(true)
+          .build();
+
+    } else {
+      return CloseableIterable.withNoopClose(snapshot.allManifests());
+    }
+  }
+
+  /**
+   * Determines the manifest files which need to be inspected because they refer to data files which
+   * can be removed after a Snapshot Expiration.
+   *
+   * @param currentTableSnapshots A list of Snapshots Currently used by the Table
+   * @param validIds              The Ids of the Snapshots which have not been expired
+   * @param expiredIds            The Ids of the Snapshots which have been expired
+   * @param currentTableMetadata  The metadata of the table being expired
+   * @param ops                   The Table Operations module for the table in question, required
+   *                              for several IO operations
+   * @return
+   */
+  public static ManifestExpirationChanges determineManifestChangesFromSnapshotExpiration(
+      List<Snapshot> currentTableSnapshots, Set<Long> validIds, Set<Long> expiredIds,
+      TableMetadata currentTableMetadata, TableOperations ops) {
+
+    Set<ManifestFile> validManifests = getValidManifests(currentTableSnapshots, ops);
+    Set<ManifestFile> manifestsToScan = validManifestsInExpiredSnapshots(validIds,
```

Comment:
As far as I remember, `validManifests` contains manifests referenced by any of current snapshot and `manifestsToScan` contains a subset of valid manifests that were written by expired snapshots but still are still referenced by some current snapshots. We have to scan such manifests if they contain deleted files so that we can remove files that were deleted by expired snapshots while keeping all files that were added by those snapshots.

This part is tricky so I think some comments would help here.

Reply:
ManifestsToScan contains entries from two different loops (This is the same as the old code)

Snapshots that survived expiration loop
1. List all Snapshots which haven't expired
2. Get every manifest from each of those snapshots
3. If that manifest is originally from an expired snapshot AND was cherrypicked or is an ancestor AND there are deletes in the manifest
   Then it's possible that those deletes can be realized and we can remove those files so we need to scan that manifest

Snapshots which did not survive Expiration and don't effect the current state
1. List all expired snapshots from the old state which have no effect on the current table state (SnapshotsNotInTableState)
2. Get all the manifests in these snapshots that we didn't scan in the above loop 
3. For these expired manifests check if they need to be scanned
  a. If they have deletes we need to Scan
  b. If they have files added we need to Scan (for Revert)

Reply:
This may be a good time to rename ManifestsToScan and Revert to, manifestsWithDeletes, manifestsWithAdds?